Provide an SMT solver callback to solc

leonardoalt opened the pull request:
Maybe this is what was intended in the first place.
If the smtchecker is used only by `solcjs --standard-json` the development frameworks still don't use it. By moving it to `compileStandardWrapper` it works out of the box (when it's released and the frameworks update their solc versions).
Example using truffle (after updating solc manually):
```
$ truffle compile
Compiling ./contracts/assert.sol...

Compilation warnings encountered:

Warning: This is a pre-release compiler version, please do not use it in production.
,/home/leonardo/devel/truffle_test/contracts/assert.sol:7:3: Warning: Assertion violation happens here
		assert(x > 0);
		^-----------^
  for:
  x = 0
```

axic commented:
I don't think this is a good idea, because the wrapper has no business in executing filesystem stuff. What I meant by having a callback for the `solver` is that callback could be provided in the `compileStandardWrapper`.

leonardoalt commented:
But who registeres the callback? Frameworks use `compileStandardWrapper` directly, so where would `handleSMTQueries` be called?

leonardoalt commented:
Or do you mean for `compileStandardWrapper` to register the callback?

axic commented:
> But who registeres the callback?

It is supplied to `compileStandardWrapper`. The node file access tools cannot be compiled for a browser environment and that means the current version of the PR breaks on Remix for example.

leonardoalt commented:
Ok, I understand now, makes sense.
But then again, who would supply the "smt callback" to `compileStandardWrapper`? Truffle, eg?

axic commented:
My new API proposal is: `compiler(standardJson, callbacks)` where `callbacks` is an object having two optional fields (`importCallback`, `smtSolverCallback`).

If the `callbacks` argument is not an object, but a function, then it is considered as `importCallback`.

Then truffle for example would have the opportunity to call `compile` with the smt solver:
```
var solc = require('solc')
var smtSolver = require('solc/smtSolver')

var output = solc.compile(input, { importCallback: resolveImports, smtSolverCallback: smtSolver.execResolver })
```

It has to be driven by the caller.

leonardoalt commented:
@axic I changed this PR to the new API you proposed. Moreover, on the Solidity side, added a new callback to be used by `SmtLib2Interface` as a sync SMT solver (https://github.com/ethereum/solidity/pull/5642), to solve everything in one run. If the callback is not given, it still uses the JSON queries with the double run (if standard JSON).
In this PR `solcjs` gives `solc` the `smtCallback` by default.
Frameworks can call the compiler in a similar way if they run locally.

leonardoalt commented:
This PR should only be merged iff https://github.com/ethereum/solidity/pull/5642 is merged.

axic commented:
Why is the Solidity change required? My proposal didn't require that change :)

leonardoalt commented:
I know, I talked with @chriseth and @ekpyron and we decided to give the single run a go, so I extended your proposal for that :p

axic commented:
We had this double callback system initially (months ago) and decided it wasn't the best.

axic commented on solcjs:
```diff
@@ -54,29 +55,26 @@ function abort (msg) {
+      'component': 'general',
+      'formattedMessage': e.toString(),
+      'message': e.toString(),
+      'type': 'Warning'
```
Keys don't need to be quoted in javascript.

axic commented on wrapper.js:
```diff
@@ -40,54 +51,65 @@ function setupMethods (soljson) {
   var compileJSON = null;
   if ('_compileJSON' in soljson) {
-    compileJSON = soljson.cwrap('compileJSON', 'string', ['string', 'number']);
+    compileJSON = soljson.cwrap('compileJSON', 'string', ['string', 'number', 'number']);
```
These old APIs cannot be changed. It breaks backwards compatibility.

axic commented on wrapper.js:
```diff
@@ -40,54 +51,65 @@ function setupMethods (soljson) {
   }
   if ('_solidity_compile' in soljson) {
-    var solidityCompile = soljson.cwrap('solidity_compile', 'string', ['string', 'number']);
+    var solidityCompile = soljson.cwrap('solidity_compile', 'string', ['string', 'number', 'number']);
```
We cannot just change this. We need to define a new function name. This breaks backwards compatibility. There doesn't seem to be a way to detect the signature of a given symbol. Well we could detect the signature based on what `solidity_version` returns.

leonardoalt commented:
Ok, before we go on with reviewing, we should decide then if want this or not.

axic commented:
This new version of the PR doesn't actually support what I proposed. There is no way to use local solver as a callback.

leonardoalt commented:
I don't understand why not. It's working fine with my local runs. Of course it won't work on browsers, but with frameworks that run locally it should, shouldn't it?

leonardoalt commented:
https://github.com/ethereum/solidity/pull/5675

axic commented on smtchecker.js:
```diff
@@ -6,7 +6,7 @@ var tmp = require('tmp');
   {
     name: 'z3',
-    params: ''
+    params: '-smt2'
```
Is this a separate fix?

axic commented on smtchecker.js:
```diff
@@ -16,17 +16,37 @@ var potentialSolvers = [
 
 function solve (query) {
-  var tmpFile = tmp.fileSync();
+  var tmpFile = tmp.fileSync({ postfix: '.smt2' });
```
Is this a separate fix?

axic commented on smtchecker.js:
```diff
@@ -16,17 +16,37 @@ var potentialSolvers = [
+  try {
+    var solverOutput = execSync(
+      solvers[0].name + ' ' + solvers[0].params + ' ' + tmpFile.name, {
+        timeout: 10000
```
Is this a separate fix?

axic commented on smtchecker.js:
```diff
@@ -16,17 +16,37 @@ var potentialSolvers = [
+      !outputStr.startsWith('unsat') &&
+      !outputStr.startsWith('unknown')
+    ) {
+      throw new Error('Failed solve SMT query. ' + e.toString());
```
Is this a separate fix?

leonardoalt replied:
This yes, the others are rather enhancements.

axic replied:
Can you create a PR for these fixes? Perhaps on top of #325.

leonardoalt replied:
https://github.com/ethereum/solc-js/pull/328

axic commented:
Removed the "z3 timeout" change from this. It still needs to be rebased if we want to continue with this.

axic commented:
@leonardoalt can this be rebased? is it worth rebasing?

leonardoalt commented:
We can close it.

elenadimitrova commented:
Ability to easily compile contracts with smt solver is a major barrier to entry. I don't understand why this is being closed is there a replacement issue to track this gap?

chriseth commented:
@elenadimitrova https://github.com/ethereum/solidity/pull/5675

chriseth commented:
@elenadimitrova perhaps to clarify: This is a change in the API and thus we need to wait for the next breaking release.

leonardoalt commented:
I'm working on this again.
https://github.com/ethereum/solidity/pull/7447

leonardoalt commented:
Too old to rebase.
Closing in favor of https://github.com/ethereum/solc-js/pull/382
